File: app/src/main/java/com/example/merchio/adapters/ProductAdapter.java
package com.example.merchio.adapters;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.merchio.R;
import com.example.merchio.models.Product;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ProductViewHolder> {

    public interface OnProductClickListener {
        void onProductClick(Product product);
        void onAddToCartClick(Product product);
    }

    private final Context context;
    private final List<Product> products;
    private final OnProductClickListener listener;

    public ProductAdapter(Context context, List<Product> products, OnProductClickListener listener) {
        this.context = context;
        this.products = new ArrayList<>(products);
        this.listener = listener;
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_product_search, parent, false);
        return new ProductViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        Product product = products.get(position);

        holder.txtCategory.setText(product.getCategoryName());
        holder.txtName.setText(product.getName());
        holder.txtPrice.setText(formatRupiah(product.getPrice()));

        if (!TextUtils.isEmpty(product.getImageUrl())) {
            Glide.with(context)
                    .load(product.getImageUrl())
                    .into(holder.imgProduct);
        } else {
            holder.imgProduct.setImageDrawable(null);
        }

        holder.itemView.setOnClickListener(v -> {
            if (listener != null) {
                listener.onProductClick(product);
            }
        });

        holder.btnCart.setOnClickListener(v -> {
            if (listener != null) {
                listener.onAddToCartClick(product);
            }
        });
    }

    @Override
    public int getItemCount() {
        return products.size();
    }

    public void updateList(List<Product> newProducts) {
        products.clear();
        products.addAll(newProducts);
        notifyDataSetChanged();
    }

    private String formatRupiah(int price) {
        NumberFormat formatter = NumberFormat.getCurrencyInstance(new Locale("id", "ID"));
        return formatter.format(price).replace(",00", "");
    }

    static class ProductViewHolder extends RecyclerView.ViewHolder {

        ImageView imgProduct, btnCart;
        TextView txtCategory, txtName, txtPrice;

        public ProductViewHolder(@NonNull View itemView) {
            super(itemView);

            imgProduct = itemView.findViewById(R.id.img_product);
            btnCart = itemView.findViewById(R.id.btn_cart);
            txtCategory = itemView.findViewById(R.id.txt_category);
            txtName = itemView.findViewById(R.id.txt_name);
            txtPrice = itemView.findViewById(R.id.txt_price);
        }
    }
}

File: app/src/main/java/com/example/merchio/adapters/ProductAdapter.kt
package com.example.merchio.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.merchio.R
import com.example.merchio.data.model.ProductDto
import com.example.merchio.utils.CurrencyFormatter
import com.example.merchio.utils.loadMerchioImage

class ProductAdapter(
    private val onClick: (ProductDto) -> Unit,
    private val onCartClick: (ProductDto) -> Unit
) : RecyclerView.Adapter<ProductAdapter.ProductViewHolder>() {

    private val items = mutableListOf<ProductDto>()

    fun submitList(data: List<ProductDto>) {
        items.clear()
        items.addAll(data)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_product_search, parent, false)
        return ProductViewHolder(view)
    }

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) = holder.bind(items[position])
    override fun getItemCount(): Int = items.size

    inner class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val image: ImageView = itemView.findViewById(R.id.img_product)
        private val category: TextView = itemView.findViewById(R.id.txt_category)
        private val name: TextView = itemView.findViewById(R.id.txt_name)
        private val price: TextView = itemView.findViewById(R.id.txt_price)
        private val cart: View = itemView.findViewById(R.id.btn_cart)

        fun bind(item: ProductDto) {
            image.loadMerchioImage(item.imageUrl)
            category.text = if (item.stock > 0) "Stock ${item.stock}" else "Out of stock"
            name.text = item.name
            price.text = CurrencyFormatter.rupiah(item.price)
            itemView.setOnClickListener { onClick(item) }
            cart.setOnClickListener { onCartClick(item) }
        }
    }
}
